Add support for repeated XML elements without a name attribute

amoerie opened the pull request:
New PR because #43722 was out of date and CI was broken at the time. See that PR for initial review. 

Fixes #36541
Fixes #36561

This pull request adds support in Microsoft.Extensions.Configuration.Xml for repeated XML elements without requiring a Name attribute.
This solves a particularly subtle bug when configuring Serilog from an XML configuration source. For a full description, see #36541

The original implementation of the XmlStreamConfigurationProvider has been modified to do the following:

Maintain a stack of encountered XML elements while traversing the XML source. This is needed to detect siblings.
When siblings are detected, automatically append an index to the generated configuration keys. This makes it work exactly the same as the JSON configuration provider with JSON arrays.
Tests are updated to reflect the new behavior:

Tests that verified an exception occurs when entering duplicate keys have been removed. Duplicate keys are supported now.
Add tests that verify duplicate keys result in the correct configuration keys, with all the lower/upper case variants and with support for the special "Name" attribute handling.
Thank you for your consideration.

amoerie committed:
Add support for repeated XML elements without a name attribute

This commit adds support in Microsoft.Extensions.Configuration.Xml for repeated XML elements without requiring a Name attribute.
This solves a particularly subtle bug when configuring Serilog from an XML configuration source. For a full description, see #36541

The original implementation of the XmlStreamConfigurationProvider has been modified to do the following:

- Maintain a stack of encountered XML elements while traversing the XML source. This is needed to detect siblings.
- When siblings are detected, automatically append an index to the generated configuration keys. This makes it work exactly the same as the JSON configuration provider with JSON arrays.

Tests are updated to reflect the new behavior:
- Tests that verified an exception occurs when entering duplicate keys have been removed. Duplicate keys are supported now.
- Add tests that verify duplicate keys result in the correct configuration keys, with all the lower/upper case variants and with support for the special "Name" attribute handling.

maryamariyan commented:
Configuration tests failing shown in https://dev.azure.com/dnceng/public/_build/results?buildId=884531&view=ms.vss-test-web.build-test-results-tab

amoerie committed:
Disable tests for duplicate handling in XML config, they are incompatible with new implementation

amoerie committed:
Stop using the 'Name' parameter to use arrays in XML in the tests, this is no longer required

amoerie commented:
@maryamariyan CI is green now. The tests that verify the configuration with duplicate keys in them are not compatible with the way the repeated elements work in XML configuration. 

To be more specific, the 'DuplicatesTestConfig' provides the following test data:

```csharp
public static TestSection DuplicatesTestConfig { get; }
                = new TestSection
                {
                    Values = new[]
                    {
                        ("Key1", (TestKeyValue)"Value1"),
                        ("Key1", (TestKeyValue)"Value1")
                    },
                    ...
                };
```

Later, this is verified:

```csharp
protected virtual void AssertConfig(
            IConfigurationRoot config,
            bool expectNulls = false,
            string nullValue = null)
        {
            var value1 = expectNulls ? nullValue : "Value1";
            ...

            Assert.Equal(value1, config["Key1"], StringComparer.InvariantCultureIgnoreCase);
            ...
        }
```

But this fails, because "Key1" is not in the output. This is because the XML config provider will produce two entries, "Key1:0" and "Key1:1", both with value "Value1". 
This is not a breaking change, because previously the XML provider simply threw an exception in this scenario. 
However, it does mean we cannot simply reuse the duplicates tests from `ConfigurationProviderTestBase`, because  the output does not match.

amoerie commented:
Any update on this?

maryamariyan commented:
> Any update on this?

We'll be reviewing this week, 
cc: @safern

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/tests/XmlConfigurationTest.cs:
```diff
@@ -193,6 +219,125 @@ public void NameAttributeInRootElementContributesToPrefix()
+                      <ConnectionString>TestConnectionString1</ConnectionString>
+                      <ConnectionString>TestConnectionString2</ConnectionString>
+                  </DefaultConnection>
+                  <DefaultConnection>
```
What would happen if I have:
```xml
<settings>
    <DefaultConnection Name='Data'>
        <ConnectionString>TestConnectionString1</ConnectionString>
        <Provider>SqlClient1</Provider>
    </DefaultConnection>
    <DefaultConnection Name='Data'>
        <ConnectionString>TestConnectionString2</ConnectionString>
        <Provider>SqlClient2</Provider>
    </DefaultConnection>
    <OtherValue>
        <Value>MyValue</Value>
    </OtherValue>
    <DefaultConnection>
        <ConnectionString>TestConnectionString3</ConnectionString>
        <Provider>SqlClient3</Provider>
    </DefaultConnection>
</settings>
```

Would I get that configuration section with: `xmlConfigSrc.Get("DefaultConnection:2:ConnectionString")` or with `xmlConfigSrc.Get("DefaultConnection:3:ConnectionString")` ? Is the index the index for the element, or is the index based on the number of elements with the same key name?

maryamariyan replied:
For this example it would return `TestConnectionString3`. The following logic added in the src:

```c#
                            // check if this element has appeared before
                            XmlConfigurationElement sibling = allElements.Where(e => e.IsSibling(element)).OrderByDescending(e => e.Index).FirstOrDefault();
                            if (sibling != null)
                            {
                                sibling.Multiple = element.Multiple = true;
                                element.Index = sibling.Index + 1;
                            }
```
keeps track of repeated elements with the same key, with Multiple/Index values having default values of false/0, and elements with the same key name increment the index.

amoerie replied:
@safern 

@maryamariyan is correct, it returns indices in line with what the equivalent JSON would produce. The index is based on elements with the same key. So in your example, you would need `xmlConfigSrc.Get("DefaultConnection:2:ConnectionString")`

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/IXmlConfigurationValue.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        string Key { get; }
+        string Value { get; }
+        string LineInfo { get; }
```
Why do we need LineInfo?

amoerie replied:
Good question. The previous implementation provided this, and I just made an effort to retain the old behavior.

maryamariyan replied:
As-is now, `GetLineInfo(reader)` seems to only get called when throwing FormatException.

I am not yet sure about the benefits of the refactoring done within the PR. (which introduces a new interface, and new classes). Could we keep the same code flow with minimal refactor while doing the fix?

amoerie replied:
The main reason for the extra classes is keeping track of encountered XML elements and their path from the root, to detect siblings. 

The old implementation was very lean and did not keep track of anything special aside from the path back to the root, but to detect siblings you must know which keys came before. Not just that, the Name behavior complicates things. 

Different combinations of elements with or without names can produce identical keys, but would not necessarily be siblings.

There might be a leaner way of implementing the fix: walking through the contents in breadth-first manner (currently the implementation is depth first) but that would - IMHO - be an equally sizeable refactor to the current version.

amoerie replied:
@maryamariyan 

Update: I've made the process leaner and simpler by implementing a few changes:

1. Only look for siblings in children of the current parent. The previous implementation searched through all encountered elements.
2. Generate key values at the end, after traversal. This allowed me to simplify a lot of the logic. 

Unfortunately I was not able to remove the new classes, like you asked. They are still needed because we need to keep track of siblings. If a new element is encountered during traversal that has the same name as a previous element (at the same depth), this means they are siblings. And when an element has siblings, that changes its configuration key.

This means we can no longer produce the configuration keys + values during traversal, but we must do so after traversal has completed. If you have suggestions on how to further trim my additions, I'll be interested to hear them.

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationElement.cs:
```diff
@@ -0,0 +1,71 @@
+                if (Multiple) tokens.Add(Index.ToString());
+
+                // the root element without a name attribute does not contribute to prefix at all
+                if (!tokens.Any()) return null;
```
Why instantiate a `tokens` list every time this Key getter block gets called?
Also, it seems like Parent, ElementName and Name are readonly fields, so won't need to check them each time this gets called.

amoerie replied:
@maryamariyan Please see my comment here: https://github.com/dotnet/runtime/pull/44608#issuecomment-748182186

I believe this comment is now outdated.

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationElementContent.cs:
```diff
@@ -0,0 +1,33 @@
+        {
+            Value = content ?? throw new ArgumentNullException(nameof(content));
+            LineInfo = lineInfo ?? throw new ArgumentNullException(nameof(lineInfo));
+            _elementPath = elementPath?.Reverse().ToArray() ?? throw new ArgumentNullException(nameof(elementPath));
```
Perhaps you could only save the keys since that is all you need in line 27 (for Key)

amoerie replied:
@maryamariyan Please see my comment here: https://github.com/dotnet/runtime/pull/44608#issuecomment-748182186

I believe this comment is now outdated.

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.Xml/tests/ConfigurationProviderXmlTest.cs:
```diff
@@ -33,11 +43,8 @@ public override void Has_debug_view()
     Key3=Value123 ({providerTag})
     Key3a:
       0=ArrayValue0 ({providerTag})
-        Name=0 ({providerTag})
```
Why would this behavior need to change?

amoerie replied:
The debug view test (ab)used the "Name" attribute for arrays. 
This is no longer necessary since the new implementation can handle arrays without having to use the "Name" attribute.

amoerie committed:
Merge branch 'master' of github.com:dotnet/runtime into repeatedelementsinxmlconfig

amoerie committed:
Rework the XML configuration provider

The XML configuration provider now builds a simple in-memory model of the XML elements that it encounters
Each element keeps track of its attributes, children and content.
Furthermore, each element also has a reference to its siblings.
Each group of siblings all share the same list.

All of the above makes it possible to intelligently produce configuration keys and values, taking into account repeated XML elements.

amoerie commented:
@maryamariyan I took a look at your comments, and almost all of them stem from the same issue: the "Key" property was lazily computed in various places, because you cannot compute it once in the constructor. This is because the value of "IsMultiple" might change, because a new element is encountered during traversal that shares the same name. This triggers the new behavior where the "sibling index" is then added to the key.

To address your concerns, I slightly reworked the implementation, and I would be interested to hear your thoughts.

I've removed all "Key" and "Value" computation logic from the new classes, they are now extremely simple POCOs. Instead, all of the heavy lifting now happens in `XmlStreamConfigurationProvider`. 

The "algorithm" now works as follows:

1. Traverse the XML document using XmlReader. While traversing, build an object tree of XmlConfigurationElement. Each element has a reference to its attributes, text content and children. This traversal happens in `XmlStreamConfigurationProvider.Read`.
2. While traversing, detect siblings. Siblings are elements that share the same parent and element name. (If a Name attribute is present, that must also be the same)
3. After the XML document has been traversed, process the XmlConfigurationElement tree (starting from the root) and produce keys and values for the configuration dictionary. This happens in `XmlStreamConfigurationProvider.ProvideConfiguration`

amoerie committed:
Add test that verifies mixing repeated with non-repeated XML elements works as expected

amoerie commented:
@maryamariyan Any chance this PR could get another review?

maryamariyan commented:
Thanks for the update, I'll do a comprehensive review during next week,

maryamariyan commented:
Hey @amoerie, I haven't forgotten about this one, I'll make sure this keeps moving and totally appreciate how you kept the fix up to date.

amoerie commented:
Thanks! On the bright side, I guess I can say I've been contributing to Microsoft Open Source for 3 years already. 😉

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/tests/ConfigurationProviderXmlTest.cs:
```diff
@@ -75,7 +82,7 @@ private void SectionToXml(StringBuilder xmlBuilder, string sectionName, TestSect
                 {
                     for (var i = 0; i < tuple.Value.AsArray.Length; i++)
                     {
-                        xmlBuilder.AppendLine($"<{tuple.Key} Name=\"{i}\">{tuple.Value.AsArray[i]}</{tuple.Key}>");
```
Why is this changing? Is this a breaking change?

amoerie replied:
In the past, the 'Name' attribute was required to support arrays. Now, this is no longer required. I felt it was more appropriate to use the simplest solution for the test suite instead of depending on a workaround that is no longer required.

To be clear, the 'Name' attribute is still completely supported, so I don't consider it a breaking change.

safern replied:
Sounds good. Do we want to add a test that includes the `Name` attribute just to make sure we don't regress that support on the future?

amoerie replied:
I believe it should already be covered by the existing tests, but I'll add a test that explicitly tests that the Name attribute can be used to simulate arrays.

amoerie replied:
Done

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationElement.cs:
```diff
@@ -0,0 +1,55 @@
+            ElementName = elementName ?? throw new ArgumentNullException(nameof(elementName));
+            Name = name;
+            LineInfo = lineInfo;
+            Children = null;
```
all these would already be null, right?

amoerie replied:
There is no need indeed, but I generally like to be explicit about it. I'll remove those lines since they're not the norm in this repo.

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationElement.cs:
```diff
@@ -0,0 +1,55 @@
+                throw new ArgumentNullException(nameof(xmlConfigurationElement));
+            }
+
+            return string.Equals(ElementName, xmlConfigurationElement.ElementName)
```
Should this comparison be ordinal or culture aware?

amoerie replied:
I think ordinal is the correct choice here. For example, the German 'ß' character is equal to 'ss' when comparing with invariant, but not when comparing with ordinal. And 'ß' is definitely not always the same as 'ss'. 
(I'm Belgian, but I've studied German language)

According to https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-5.0, configuration keys are case insensitive. I'm thinking that should also mean siblings should be compare case insensitively, since they would end up as the same keys eventually anyway. 

I'll update the calls.

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs:
```diff
@@ -40,61 +39,93 @@ public class XmlStreamConfigurationProvider : StreamConfigurationProvider
                             {
-                                prefixStack.Pop();
+                                var element = new XmlConfigurationElement(reader.LocalName, GetName(reader), GetLineInfo(reader));
+
```
remove empty line.

amoerie replied:
On it

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationElement.cs:
```diff
@@ -0,0 +1,55 @@
+
+        public bool IsSiblingOf(XmlConfigurationElement xmlConfigurationElement)
+        {
+            if (xmlConfigurationElement is null)
```
@stephentoub do you know if we prefer `== null` or `is null`?

stephentoub replied:
is null is fine

amoerie replied:
👍

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs:
```diff
@@ -161,49 +184,152 @@ private static string GetLineInfo(XmlReader reader)
+
+            return configuration;
+
+            void ProcessElement(List<string> prefix, XmlConfigurationElement element)
```
Any reason to make this local methods instead of private static?

amoerie replied:
Two reasons:
1. It allows me to not have to pass around the `Dictionary<string, string> configuration` variable. 
2. These methods conceptually belong to the ProvideConfiguration method, as standalone methods they are not really reusable. They only make sense in the context of the higher level ProvideConfiguration method.

If you prefer private static methods I can change it, obviously.

safern replied:
Sounds good. I was curious and missed the fact that you would need to pass the `configuration` dictionary around.

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs:
```diff
@@ -40,61 +39,93 @@ public class XmlStreamConfigurationProvider : StreamConfigurationProvider
+
+                                XmlConfigurationElement parent = currentPath.Any() ? currentPath.Peek() : null;
+
+                                if (parent == null)
```
We could avoid this check right? We could do something like:

```cs
if (!currentPath.Any())
{
    root = element;
}
else
{
     XmlConfigurationElement parent = currentPath.Peek();
     ...
}
```

I don't believe `currentPath.Peek()`can return null right?

amoerie replied:
Nice find, updating.

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationElementTextContent.cs:
```diff
@@ -0,0 +1,20 @@
+
+        public string TextContent { get; }
+
+        public string LineInfo { get; }
```
Do we need to store `LineInfo`? I don't see it used other than for exceptions and it is passed directly from `GetLineInfo(reader)` call as opposed from here.

amoerie replied:
I think we need it. In XmlStreamConfigurationProvider#306 the line info is passed to AddToConfiguration. In that method, a duplicate key exception can be thrown which needs the line info. 

At this point in time, the XML reader has finished its work completely, so we can no longer retrieve the line info from the reader.

safern replied:
Yup I missed that part. Disregard this comment.

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/tests/ConfigurationProviderXmlTest.cs:
```diff
@@ -20,6 +20,16 @@ public override void Combine_after_other_provider()
 
+        public override void Load_from_single_provider_with_duplicates_throws()
+        {
+            // Disabled test due to XML handling of duplicate keys section.
```
Why not just remove those tests from the base? 

Is this a breaking change that we should put behind a flag?

amoerie replied:
> Why not just remove those tests from the base?

These implementations disable/override the test:

- ConfigurationProviderCommandLineTest
- ConfigurationProviderEnvironmentVariablesTest
- ConfigurationProviderIniTest
- ConfigurationProviderJsonTest
- **ConfigurationProviderXmlTest** (since this PR)

These implementations don't override it:

ConfigurationProviderMemoryTest
ConfigurationProviderIniTest

So it looks like there's still two providers that throw an exception for duplicate keys within a single provider.

> Is this a breaking change that we should put behind a flag?

The only way in which this is breaking, is that it no longer throws an exception where it previously did. If that is considered a breaking change, then maybe a flag is appropriate. IMHO I don't think it should be a flag.

safern replied:
> So it looks like there's still two providers that throw an exception for duplicate keys within a single provider.

Ah ok, I missed the fact that there was still two providers that threw an exception. 

> If that is considered a breaking change.

I confirmed and we don't consider that a breaking change.

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationElement.cs:
```diff
@@ -0,0 +1,55 @@
+        /// <summary>
+        /// The children of this element
+        /// </summary>
+        public List<XmlConfigurationElement>? Children { get; set; }
```
It seems like we haven't enabled nullable for this project yet.

amoerie replied:
OK I'll remove the annotations

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs:
```diff
@@ -40,61 +39,93 @@ public class XmlStreamConfigurationProvider : StreamConfigurationProvider
-                                prefixStack.Pop();
+                                var element = new XmlConfigurationElement(reader.LocalName, GetName(reader), GetLineInfo(reader));
+
+                                XmlConfigurationElement parent = currentPath.Any() ? currentPath.Peek() : null;
```
Let's use `currentPath.Count() != 0`, there is a PR removing LINQ usages for configuration: https://github.com/dotnet/runtime/pull/47873/files#diff-aae949242d873fc3ce6c6e560cc62149d3ecb5f65929936795a812bbc2f950a7R71

stephentoub replied:
> Let's use currentPath.Count() != 0

Count() is LINQ. You want Count, the property.

safern replied:
Yup, you're right. Typing without thinking.

amoerie replied:
I'll remove System.Linq and fix any usages

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs:
```diff
@@ -40,61 +39,93 @@ public class XmlStreamConfigurationProvider : StreamConfigurationProvider
+                                    else
+                                    {
+                                        // check if this element has appeared before, elements are considered siblings if their element names match
+                                        XmlConfigurationElement sibling = parent.Children.FirstOrDefault(e => element.IsSiblingOf(e));
```
Let's avoid using LINQ, for perf reasons there is a PR removing LINQ usages here: https://github.com/dotnet/runtime/pull/47873/files#diff-aae949242d873fc3ce6c6e560cc62149d3ecb5f65929936795a812bbc2f950a7R71

stephentoub replied:
Count() is LINQ. You want Count, the property.

amoerie replied:
I've added a method XmlConfigurationElement.FindSiblingInChildren that uses a simple for loop.
This also allowed me to make IsSiblingOf a private method.

amoerie committed:
Cleanup nullable annotations in XmlConfigurationElement

Also don't explicitly initialize properties to null

amoerie committed:
Use ordinal ignore case when detecting siblings in XML configuration

amoerie committed:
Remove empty line

amoerie committed:
Remove usage of Linq method '.Any()'

amoerie committed:
Remove dependency on System.Linq in XmlStreamConfigurationProvider

amoerie committed:
Simplify check that detects whether the current element is the root element

amoerie commented:
I've pushed updates that address the feedback. Thanks for the extensive review @safern!

amoerie committed:
Merge branch 'master' of github.com:dotnet/runtime into repeatedelementsinxmlconfig

safern commented:
> Edit: looks like CI isn't happy again.

Yeah, CI was broken because of: https://github.com/dotnet/install-scripts/issues/136

I will close and re-open to trigger it again.

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationElement.cs:
```diff
@@ -0,0 +1,65 @@
+
+        public XmlConfigurationElement FindSiblingInChildren(XmlConfigurationElement element)
+        {
+            if (element is null) throw new ArgumentNullException(nameof(element));
```
```suggestion
            if (element is null)
                throw new ArgumentNullException(nameof(element));
```

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/tests/XmlConfigurationTest.cs:
```diff
@@ -193,6 +219,159 @@ public void NameAttributeInRootElementContributesToPrefix()
+        }
+
+        [Fact]
+        public void NestedRepeatedElementsContributeToPrefix()
```
Should we add non-case sensitive tests per:  https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-5.0

amoerie replied:
Yes, to my surprise I couldn't find any case sensitivity related tests. I'll add a few scenarios.

amoerie replied:
Added three tests:

1. Simple config with mixed casing works
2. Repeated XML elements with mixed casing are properly recognized as siblings
3. Duplicate keys with mixed casings throw an exception

safern replied:
Thank you 😄

amoerie committed:
Apply suggestions from code review

Co-authored-by: Santiago Fernandez Madero <safern@microsoft.com>

amoerie committed:
Add test for array simulation using Name attribute

amoerie committed:
Add tests related to case insensitivity in XML configuration keys

Verify that a duplicate key exception is thrown when keys are duplicate with different casing
Verify that XML siblings are properly detected when the siblings have different casing
Verify that values with keys that were originally upper case can be retrieved with their lower case counterparts

amoerie committed:
Merge master and resolve conflicts

amoerie commented:
CI is mostly green now, but still a few checks are failing. At first sight it doesn't seem related to what I'm doing here, but I'll let you be the judge on that..

safern reviewed (dismissed):
LGTM

safern commented:
@amoerie can we do some benchmarking to measure how this change impacts perf?

amoerie commented:
> @amoerie can we do some benchmarking to measure how this change impacts perf? 

Sure, do you know if any existing benchmarks exist? If not, is there a guideline / template / boilerplate project to add new benchmarks? Or do I just create a throwaway benchmark project? 

Without even benchmarking, I can tell you that we will have a little more memory allocation. This is because we're now building an object tree that represents the XML document, unlike before, where the implementation would immediately build the key value dictionary as it walked through the XML. This extra memory allocation will grow linearly with the size of the input. 

Speed wise I also expect the new implementation to be a little slower, mostly because we're looking for siblings after each new element. With XML documents where one parent node has thousands of child nodes of different types, this might become painful. 

Then again, all of this must be put into context: in usual circumstances, this code only runs on startup, and probably won't handle XML files that are multiple megabytes in size.

amoerie commented:
Update: because I don't really know how to add a new benchmark project here, I've created a little standalone repo for the benchmarks: https://github.com/amoerie/dotnet-runtime-benchmarks

As the baseline, I'm using the Microsoft.Extensions.Configuration.Xml nuget package. For the changes, I've copied the new source code to that repository. This allows me to have both implementations side by side. 

Then, I added benchmarks for four scenarios:
1. simple.xml - an appsettings file with 50 simple key values
2. deep.xml - an appsettings file with 1 key value that is 50 XML elements deep
3. names.xml - an appsettings file that simulates an array using the Name attribute
4. repeated.xml - same as names.xml but without the Name attribute. This crashes on the old implementation, but you can compare these results with names.xml

These are the results:

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.103
  [Host]     : .NET Core 5.0.3 (CoreCLR 5.0.321.7212, CoreFX 5.0.321.7212), X64 RyuJIT
  DefaultJob : .NET Core 5.0.3 (CoreCLR 5.0.321.7212, CoreFX 5.0.321.7212), X64 RyuJIT


```
| Method |     TestFile |      Mean |    Error |   StdDev | Ratio | RatioSD |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|------- |------------- |----------:|---------:|---------:|------:|--------:|--------:|-------:|------:|----------:|
| **Before** |     **deep.xml** | **121.67 μs** | **1.902 μs** | **1.779 μs** |  **1.00** |    **0.00** | **14.6484** | **1.9531** |     **-** |   **92048 B** |
|  After |     deep.xml | 129.19 μs | 1.353 μs | 1.265 μs |  1.06 |    0.02 | 15.8691 | 1.9531 |     - |   99800 B |
|        |              |           |          |          |       |         |         |        |       |           |
| **Before** |    **names.xml** | **188.59 μs** | **3.553 μs** | **3.489 μs** |  **1.00** |    **0.00** | **24.1699** | **4.6387** |     **-** |  **151912 B** |
|  After |    names.xml | 221.53 μs | 3.172 μs | 2.967 μs |  1.17 |    0.02 | 27.3438 | 6.1035 |     - |  172088 B |
|        |              |           |          |          |       |         |         |        |       |           |
| **Before** | **repeated.xml** |        **NA** |       **NA** |       **NA** |     **?** |       **?** |       **-** |      **-** |     **-** |         **-** |
|  After | repeated.xml | 182.69 μs | 1.858 μs | 1.738 μs |     ? |       ? | 22.9492 | 4.1504 |     - |  144720 B |
|        |              |           |          |          |       |         |         |        |       |           |
| **Before** |   **simple.xml** |  **92.82 μs** | **1.590 μs** | **1.409 μs** |  **1.00** |    **0.00** | **11.5967** | **1.2207** |     **-** |   **72864 B** |
|  After |   simple.xml | 110.03 μs | 1.846 μs | 1.727 μs |  1.18 |    0.03 | 12.5732 | 1.3428 |     - |   79096 B |

Benchmarks with issues:
  XmlConfigurationBenchmarks.Before: DefaultJob [TestFile=repeated.xml]

Summary:
- The new implementation is somewhere between 5% and 20% slower relative to the previous one. 
- The new implementation allocates between 5% and 15% more relative to the previous one.

amoerie commented:
@maryamariyan Any chance this can get some feedback again?

safern commented:
@amoerie thanks for running the benchmarks. I don't think we have any benchmarks for Microsoft.Extensions.Configuration yet: https://github.com/dotnet/performance/tree/master/src/benchmarks/micro/libraries

based on the results and your investigation, I think taking a 5-20% regression is not acceptable, specially since we could try and improve your current implementation with the info you got from the profiling. 

> GetLineInfo : previously, line info was only read when an error occurred. Now, we must call it for every element and store it, in case we need to throw an exception later. One might question if this line info is really such an important feature.

Is there a way to get the line info only if an exception is thrown? Removing it from the error might be breaking and I think having it in the error is valuable as it helps you know where the error in your configuration file is.

amoerie committed:
Improve performance of XML configuration provider

amoerie commented:
@safern I've pushed some optimizations, if you would find the time again to review that.

> > GetLineInfo : previously, line info was only read when an error occurred. Now, we must call it for every element and store it, in case we need to throw an exception later. One might question if this line info is really such an important feature.
> 
> Is there a way to get the line info only if an exception is thrown? Removing it from the error might be breaking and I think having it in the error is valuable as it helps you know where the error in your configuration file is.

Avoiding the line info reading would require completely revising the algorithm, which I'm hesitant to do right now. But I found that the most time was spent on `SR.Format` which uses the `LineNumber` and `LinePosition`, so now I'm delaying that string format as long as possible.

This is a summary of the improvements I've made:

- Store LineNumber and LinePosition instead of LineInfo. This avoids a call to string.Format in most cases.
- Use a helper class to build the configuration keys. This uses `StringBuilder `instead of `string.Join` internally, which is much faster.
- Keep a dictionary of XML children, keyed by their "SiblingName" (= ElementName + Name). This allows for a much more efficient lookup of siblings under the current parent.
- Do not create this dictionary for XML elements with a single child. This is quite common, and avoiding it in this scenario is more efficient in terms of memory and execution. 

These are the performance results of these modifications:

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.103
  [Host]     : .NET Core 5.0.3 (CoreCLR 5.0.321.7212, CoreFX 5.0.321.7212), X64 RyuJIT
  DefaultJob : .NET Core 5.0.3 (CoreCLR 5.0.321.7212, CoreFX 5.0.321.7212), X64 RyuJIT


```
|   Method |     TestFile |      Mean |    Error |   StdDev | Ratio | RatioSD |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|--------- |------------- |----------:|---------:|---------:|------:|--------:|--------:|-------:|------:|----------:|
|   **Before** |     **deep.xml** | **125.49 μs** | **2.507 μs** | **2.682 μs** |  **1.00** |    **0.00** | **14.6484** | **1.9531** |     **-** |   **92048 B** |
|    After |     deep.xml | 130.67 μs | 1.721 μs | 1.609 μs |  1.04 |    0.03 | 15.8691 | 1.9531 |     - |   99800 B |
| Improved |     deep.xml | 124.07 μs | 2.414 μs | 2.965 μs |  0.99 |    0.03 | 14.4043 | 1.8311 |     - |   90648 B |
|          |              |           |          |          |       |         |         |        |       |           |
|   **Before** |    **names.xml** | **185.49 μs** | **2.378 μs** | **2.442 μs** |  **1.00** |    **0.00** | **24.1699** | **4.6387** |     **-** |  **151912 B** |
|    After |    names.xml | 231.85 μs | 4.546 μs | 7.962 μs |  1.27 |    0.05 | 27.3438 | 6.1035 |     - |  172088 B |
| Improved |    names.xml | 181.91 μs | 1.792 μs | 1.399 μs |  0.98 |    0.02 | 25.8789 | 5.6152 |     - |  163224 B |
|          |              |           |          |          |       |         |         |        |       |           |
|   **Before** | **repeated.xml** |        **NA** |       **NA** |       **NA** |     **?** |       **?** |       **-** |      **-** |     **-** |         **-** |
|    After | repeated.xml | 183.69 μs | 2.673 μs | 2.501 μs |     ? |       ? | 22.9492 | 4.1504 |     - |  144720 B |
| Improved | repeated.xml | 156.03 μs | 2.841 μs | 3.157 μs |     ? |       ? | 20.7520 | 4.1504 |     - |  131392 B |
|          |              |           |          |          |       |         |         |        |       |           |
|   **Before** |   **simple.xml** |  **92.73 μs** | **1.156 μs** | **0.903 μs** |  **1.00** |    **0.00** | **11.5967** | **1.2207** |     **-** |   **72864 B** |
|    After |   simple.xml | 111.59 μs | 1.996 μs | 1.770 μs |  1.21 |    0.02 | 12.5732 | 1.3428 |     - |   79096 B |
| Improved |   simple.xml |  92.49 μs | 1.412 μs | 1.179 μs |  1.00 |    0.02 | 11.7188 | 1.3428 |     - |   74232 B |

Benchmarks with issues:
  XmlConfigurationBenchmarks.Before: DefaultJob [TestFile=repeated.xml]

amoerie committed:
Revert accidental change of solution file

amoerie committed:
Merge branch 'main' into repeatedelementsinxmlconfig

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs:
```diff
@@ -40,61 +39,117 @@ public class XmlStreamConfigurationProvider : StreamConfigurationProvider
+                                // If parent already has a dictionary of children, update the collection accordingly
+                                if (parent.ChildrenBySiblingName != null)
+                                {
+                                    // check if this element has appeared before, elements are considered siblings if their SiblingName properties match
+                                    if (parent.ChildrenBySiblingName.TryGetValue(element.SiblingName, out var siblings))
+                                    {
+                                        siblings.Add(element);
+                                    }
+                                    else
+                                    {
+                                        parent.ChildrenBySiblingName.Add(element.SiblingName, new List<XmlConfigurationElement> { element });
+                                    }
```
I think this would be simpler if written as: 
```suggestion
                                    // check if this element has appeared before, elements are considered siblings if their SiblingName properties match
                                    if (!parent.ChildrenBySiblingName.TryGetValue(element.SiblingName, out var siblings))
                                    {
                                        siblings = new List<XmlConfigurationElement>();
                                        parent.ChildrenBySiblingName.Add(element.SiblingName, siblings);
                                    }
                                    siblings.Add(element);
```

amoerie replied:
Agreed, I've updated it. Thanks!

safern commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs:
```diff
@@ -129,81 +184,264 @@ public override void Load(Stream stream)
+                    // If there is a namespace attached to current attribute
+                    if (!string.IsNullOrEmpty(reader.NamespaceURI))
+                    {
+                        throw new FormatException(SR.Format(SR.Error_NamespaceIsNotSupported, GetLineInfo(reader)));
```
Why do we need to throw this exception here? It seems like it in `ProcessAttributes`.

amoerie replied:
I think - please forgive me if I'm wrong - you are confusing two methods here:

1. `private static void ProcessAttributes(XmlReader reader, XmlConfigurationElement element)`
2. `void ProcessElementAttributes(Prefix prefix, XmlConfigurationElement element)`

The first method is invoked during the parsing stage. It reads the XML attributes from the XML document and adds instances of `XmlConfigurationElementAttributeValue` to the current `XmlConfigurationElement`

The second method is invoked during the processing stage. It produces configuration key values for each attribute.

To avoid future confusion, I have renamed `ProcessAttributes` (the first method) to `ReadAttributes`.

If I misunderstood you and it was not a matter of confusion, can you please clarify why you think that exception does not belong there?

amoerie committed:
Apply suggestion from feedback: simplify children list initialization

amoerie committed:
Rename ProcessAttributes -> ReadAttributes when parsing the XML

This avoids confusion with the ProcessXYZ helper methods that run during a different stage in the algorithm

amoerie commented:
> Thanks, @amoerie. I'm happy that you were able to find a way that it even improves the perf. Now that you wrote some benchmarks, can we add them to: https://github.com/dotnet/performance/tree/master/src/benchmarks/micro/libraries ? That way our infra will catch if there are any regressions in the future.
> 
> I left a couple of comments, but overall looks good.

Thanks, I'll see if I can add the benchmarks there.

amoerie committed:
Merge branch 'main' of github.com:dotnet/runtime into repeatedelementsinxmlconfig

safern commented:
Timeout is: https://github.com/dotnet/runtime/issues/49309

safern commented:
@amoerie thanks for the effort to get this fix in 🎉

amoerie commented:
You're very welcome! It's been quite the experience, I learned quite a few things during this endeavor. 💪

Thank you (and @maryamariyan too) for taking time out of your schedules to review and discuss the work of a total stranger. From my own work experience, I know this is not trivial.

For me personally, I consider it a very big deal to have contributed to .NET itself. So thanks for making that possible. 👍

safern commented:
> For me personally, I consider it a very big deal to have contributed to .NET itself. So thanks for making that possible. 

Anytime! Feel free to browse for more up-for-grabs issues and help us make .NET better and reach out for any help if needed, we really appreciate and like the community help!

eerhardt commented:
I have created https://github.com/dotnet/docs/issues/30660 to document this breaking change.
